### PR TITLE
Enable race detector for unit testing 

### PR DESCRIPTION
**Pull request by alexanderConstantinescu**
**- What this PR does and why is it needed**

A month ago we enabled and then quickly after disabled the race detector for PRs, we created the issue: https://github.com/ovn-org/ovn-kubernetes/issues/1550 to track the reinstatement. This PR re-introduces the race detector with two additional prerequisites for that to work.

1. Now that PR https://github.com/ovn-org/ovn-kubernetes/pull/1567 merged and fixed a bunch of races in the master egress IP code, we can simply disable the race detector for the node egress IP tests, due to: https://github.com/ovn-org/ovn-kubernetes/issues/1549 which will probably never be resolved. Egress IP in local gateway mode will probably never be executed anyway, so I think it's fine to disable the race detector for that
2. There are race failures in `pod_test.go` due to the `errorCache` not being thread safe in `mock_ovn.go`, I've introduced a mutex to avoid that. 

With these two points solved (as two separate commits for easier review) we can now re-enable the race detector for our unit tests, which is the revert I've done of: https://github.com/ovn-org/ovn-kubernetes/pull/1551/commits/8803550fe702370d37ceb1344e2e8b9035c11da0

This patch consistently passes `RACE=1 make check` locally on my computer, so we should be good to go. 

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/node/gateway_init_linux_test.go [hunk 1]
```diff
@@ -34,6 +34,8 @@ import (
 	. "github.com/onsi/gomega"
 )
 
+var primaryLinkName = "primary"
```
Comment (aojea) [inline go-controller/pkg/node/gateway_init_linux_test.go:37]: should we take the opportunity to change it to a constant? or it needs to be a var?

File: go-controller/pkg/node/egressip_test.go [hunk 2]
```diff
@@ -1,3 +1,5 @@
+// +build !race
```
Comment (aojea) [inline go-controller/pkg/node/egressip_test.go:1]: heh, smart

**Comment by aojea** [comment]
/lgtm
once CI passes, can you retest?, Tim just fixed the network policy issue

File: go-controller/pkg/node/gateway_init_linux_test.go (hunk 1 above)
Comment (alexanderConstantinescu) [reply go-controller/pkg/node/gateway_init_linux_test.go:37]: It does not. I can change it and relaunch the CI to check if it passes now.

Comment (aojea) [reply go-controller/pkg/node/gateway_init_linux_test.go]: it's a nit, just realized while reviewing this, maybe is a good first issue to report?, `switch vars to constants whenever is possible`?

Comment (alexanderConstantinescu) [reply go-controller/pkg/node/gateway_init_linux_test.go]: Sure, I don't see why not.

**Comment by danwinship** [comment]
> We disabled the race detector a couple of weeks back, see:
> #1550.

I know that being able to refer to issues/PRs in commit messages is supposed to be a feature, but GitHub's implementation of it is so terrible that I feel like it should be avoided. (It re-spams the issue every time you re-push your commit, and then I believe it will spam it some more when we pull it downstream.)

File: go-controller/pkg/testing/mock_ovn.go [hunk 3]
```diff
@@ -48,6 +49,9 @@ type MockOVNClient struct {
 	// error injection
 	// keys are of the form: Table:Name:FieldType
 	errorCache map[string]error
+	// Mutex synchronizing the cache when running the
+	// tests with the race detector "on"
+	errorCacheMutex sync.Mutex
```
Comment (danwinship) [inline go-controller/pkg/testing/mock_ovn.go:54]: `cache` is not thread-safe either. While we may not _currently_ be using it from multiple threads, you might as well just add a single top-level `sync.Mutex` to `MockOVNClient` and use it for all methods.

Also, you don't need to reference the race detector specifically; if we are accessing the fields from multiple goroutines, then that's a bug whether or not we are running the race detector. The race detector is just _finding_ the bug. It's not the cause of the bug.

File: go-controller/pkg/testing/mock_ovn.go [hunk 4]
```diff
@@ -64,10 +68,11 @@ var _ goovn.Execution = &MockExecution{}
 // return a new mock client to operate on db
 func NewMockOVNClient(db string) *MockOVNClient {
 	mock := &MockOVNClient{
-		db:         db,
-		cache:      make(map[string]MockObjectCacheByName),
-		errorCache: make(map[string]error),
-		connected:  true,
+		db:              db,
+		cache:           make(map[string]MockObjectCacheByName),
+		errorCacheMutex: sync.Mutex{},
```
Comment (danwinship) [inline go-controller/pkg/testing/mock_ovn.go:73]: `sync.Mutex`es intentionally do not need to be initialized

Comment (alexanderConstantinescu) [reply go-controller/pkg/testing/mock_ovn.go]: I didn't know that. It means we could probably clean that up in some other places around our code which initializes the mutexes

**Comment by alexanderConstantinescu** [comment]
> We'll want to see a bunch of passing test runs before merging...

I agree, I'll re-run this a couple of times. However, to be clear: this _only_ enables the race detector for unit tests (and I have done that locally tens of times already)

File: go-controller/pkg/testing/mock_ovn.go (hunk 4 above)
Comment (aojea) [reply go-controller/pkg/testing/mock_ovn.go]: I think that Dan means that you don'need to initialize  `errorCacheMutex` in the constructor, just remove this line, right Dan?
xref https://golang.org/doc/effective_go.html#allocation_new

Comment (aojea) [reply go-controller/pkg/testing/mock_ovn.go]: ups, sorry, my bad, I didn't understand your comment :sweat_smile:

**Comment by danwinship** [comment]
> I agree, I'll re-run this a couple of times. However, to be clear: this _only_ enables the race detector for unit tests (and I have tested this locally tens of times already)

yeah, I just mean, last time we enabled it we ended up having to revert it shortly after because there were still too many races, and it totally destroyed our CI signal

**Comment by aojea** [comment]
we've introduced a race in the unit tests lately, I noticed it was failing more than usual and after retest here the race detector is complaining

**Comment by aojea** [comment]
the race is in the egresfirewall code :thinking:

**Comment by alexanderConstantinescu** [comment]
> the race is in the egresfirewall code

I will have a look and try to fix it, however that is strange given that we have not merged anything egress firewall related. Maybe a refactor of something gone bad?

**Comment by aojea** [comment]
didn't go deep, sorry, will try to narrow it down if have some time.
Just noticed that the unit tests are failing more than usual and remembered this PR

**Comment by alexanderConstantinescu** [comment]
> didn't go deep, sorry, will try to narrow it down if have some time.
> Just noticed that the unit tests are failing more than usual and remembered this PR

I know what the problem is with the egress firewall code. I will file a PR and fix it.

**Comment by alexanderConstantinescu** [comment]
@danwinship : last ping for the day! Could you also have a look at this again, I've re-run the PR a couple of times and it seems to be working fine with all minor fixes this PR includes. Reminder: this only adds the race detector to our unit tests, not the e2e ones.

**Comment by alexanderConstantinescu** [comment]
Obviously I spoke too soon....

**Comment by alexanderConstantinescu** [comment]
```
should skip populating egress node data for nodes that have incorrect IP address
I1019 17:40:36.093183   12412 reflector.go:181] Stopping reflector *v1.EgressIP (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/egressip_test.go:1158
I1019 17:40:36.092579   12412 watch.go:181] Stopping fake watcher.
I1019 17:40:36.093287   12412 reflector.go:181] Stopping reflector *v1.Namespace (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.092658   12412 watch.go:181] Stopping fake watcher.
I1019 17:40:36.093364   12412 reflector.go:181] Stopping reflector *v1.Service (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.092711   12412 watch.go:181] Stopping fake watcher.
I1019 17:40:36.093421   12412 reflector.go:181] Stopping reflector *v1.Pod (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.092743   12412 watch.go:181] Stopping fake watcher.
I1019 17:40:36.093477   12412 reflector.go:181] Stopping reflector *v1.NetworkPolicy (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.092681   12412 watch.go:181] Stopping fake watcher.
I1019 17:40:36.093534   12412 reflector.go:181] Stopping reflector *v1.Endpoints (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.092849   12412 watch.go:181] Stopping fake watcher.
I1019 17:40:36.093592   12412 reflector.go:181] Stopping reflector *v1beta1.CustomResourceDefinition (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.104409   12412 reflector.go:175] Starting reflector *v1beta1.CustomResourceDefinition (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.104454   12412 reflector.go:211] Listing and watching *v1beta1.CustomResourceDefinition from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.204322   12412 shared_informer.go:253] caches populated
I1019 17:40:36.204881   12412 reflector.go:175] Starting reflector *v1.Service (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.205468   12412 reflector.go:175] Starting reflector *v1.Endpoints (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.205541   12412 reflector.go:211] Listing and watching *v1.Endpoints from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.205512   12412 reflector.go:211] Listing and watching *v1.Service from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.206051   12412 reflector.go:175] Starting reflector *v1.NetworkPolicy (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.206356   12412 reflector.go:175] Starting reflector *v1.Namespace (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.207512   12412 reflector.go:211] Listing and watching *v1.Namespace from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.206719   12412 reflector.go:175] Starting reflector *v1.Node (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.207652   12412 reflector.go:211] Listing and watching *v1.Node from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.206973   12412 reflector.go:175] Starting reflector *v1.Pod (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.208057   12412 reflector.go:211] Listing and watching *v1.Pod from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.207459   12412 reflector.go:211] Listing and watching *v1.NetworkPolicy from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.307008   12412 shared_informer.go:253] caches populated
I1019 17:40:36.307088   12412 shared_informer.go:253] caches populated
I1019 17:40:36.307121   12412 shared_informer.go:253] caches populated
I1019 17:40:36.307153   12412 shared_informer.go:253] caches populated
I1019 17:40:36.307551   12412 shared_informer.go:253] caches populated
I1019 17:40:36.307815   12412 shared_informer.go:253] caches populated
I1019 17:40:36.308480   12412 reflector.go:175] Starting reflector *v1.EgressIP (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.308904   12412 reflector.go:211] Listing and watching *v1.EgressIP from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.408447   12412 shared_informer.go:253] caches populated
I1019 17:40:36.409536   12412 reflector.go:175] Starting reflector *v1.EgressFirewall (0s) from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.409638   12412 reflector.go:211] Listing and watching *v1.EgressFirewall from pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/reflector.go:125
I1019 17:40:36.509632   12412 shared_informer.go:253] caches populated
E1019 17:40:36.511534   12412 ovn.go:775] egress node initialization error: invalid CIDR address: 192.168.126.510/24
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1f3e185]

goroutine 5638 [running]:
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).addEgressNode(0xc0005792c0, 0xc0005b8900, 0x2416ee3, 0x1d)
	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/egressip.go:654 +0x185
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).WatchEgressNodes.func2(0x23db7c0, 0xc00094e300, 0x23db7c0, 0xc0005b8900)
	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/ovn/ovn.go:792 +0x26f
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/home/runner/go/pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/controller.go:225
k8s.io/client-go/tools/cache.FilteringResourceEventHandler.OnUpdate(0xc0006a43c0, 0x271d040, 0xc00000f000, 0x23db7c0, 0xc00094e300, 0x23db7c0, 0xc0005b8900)
	/home/runner/go/pkg/mod/k8s.io/client-go@v0.18.6/tools/cache/controller.go:259 +0x137
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*Handler).OnUpdate(0xc0008c9aa0, 0x23db7c0, 0xc00094e300, 0x23db7c0, 0xc0005b8900)
	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/factory/factory.go:61 +0xbd
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*informer).newFederatedQueuedHandler.func2.1.1(0xc0008c9aa0)
	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/factory/factory.go:238 +0xa9
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*informer).forEachQueuedHandler(0xc0008fc880, 0xc000075eb8)
	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/factory/factory.go:107 +0x2b9
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*informer).newFederatedQueuedHandler.func2.1(0xc00046b9e0)
	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/factory/factory.go:237 +0x6a
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*informer).processEvents(0xc0008fc880, 0xc0000a8a20, 0xc00004e1e0)
	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/factory/factory.go:174 +0xba
created by github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.newQueuedInformer
	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/factory/factory.go:365 +0x1c5
FAIL	github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn	19.714s
FAIL
make: *** [check] Error 1
```

OK, that test is new and has been pulled today onto this branch with my last rebase. However, it is interesting. It only fails with the race detector on, it should always have been failing...

Theory: I suspect the overhead of running the test with the race detector set to "on" made the test "slower" and thus detected the problem. I am pushing a fix for this now

**Comment by dave-tucker** [comment]
@alexanderConstantinescu it fails with the race detector off too. I've seen it 3 times today on misc PRs.
Hopefully we can get this PR merged ASAP.

**Comment by alexanderConstantinescu** [comment]
> @alexanderConstantinescu it fails with the race detector off too. I've seen it 3 times today on misc PRs.
> Hopefully we can get this PR merged ASAP.

Yeah, I've just seen it on your PR. I will move the fix into its own PR

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -49,7 +49,7 @@ jobs:
       run: |
         set -x
         pushd go-controller
-        make check
+           RACE=1 make check
```
Comment (dave-tucker) [inline .github/workflows/test.yml:52]: Why not leave this as `make check` and make the default behaviour in `test-go.sh` be to enable the race detector. This way we won't forget to run with RACE=1 locally and be surprised by failures in CI

**Review by dave-tucker — APPROVED** [review]
Small nit, I'd prefer this to be on by default for everyone, but LGTM

File: .github/workflows/test.yml (hunk 5 above)
Comment (alexanderConstantinescu) [reply .github/workflows/test.yml:52]: You are right. I thought about this yesterday, but got caught up in other things. I will modify that

**Comment by aojea** [comment]
/lgtm
@danwinship @trozet @dcbw can we get this in?

File: go-controller/hack/test-go.sh [hunk 6]
```diff
@@ -22,26 +22,26 @@ function testrun {
...
+    # enable go race detector
+    if [ ! -z "${RACE:-}" ]; then
+        args="-race "
+    fi
     if [[ "$USER" != root && " ${root_pkgs[@]} " =~ " $pkg " ]]; then
         testfile=$(mktemp --tmpdir ovn-test.XXXXXXXX)
         echo "sudo required for ${pkg}, compiling test to ${testfile}"
-        go test -covermode set -c "${pkg}" -o "${testfile}"
-        go_test="sudo ${testfile}"
+        if [[ ! -z "${RACE:-}" && "${pkg}" != "github.com/ovn-org/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller" ]]; then
```
Comment (trozet) [inline go-controller/hack/test-go.sh:32]: @alexanderConstantinescu you forgot the " # FIXME race detector fails with hybrid-overlay tests" comment here. I'll assume that means you are volunteering to fix that too ;)

**Review by trozet — APPROVED** [review]

**Pull request merged**